cylc/flow/task_pool.py

```diff
@@ -1055,161 +1111,159 @@ def remove_suiciding_tasks(self):
-                self.suite_db_mgr.put_update_task_outputs(itask)
-        return len(bad_items)
+            # Now spawn downstream on chosen outputs.
+            for msg in msgs:
+                try:
+                    children = itask.children[msg]
+                except KeyError:
+                    pass
+                else:
+                    for child_name, child_point in children:
+                        self.spawn(itask.tdef.name, itask.point,
+                                   child_name, child_point, msg)
 
-    def remove_tasks(self, items, spawn=False):
-        """Remove tasks from pool."""
+    def remove_tasks(self, items):
+        """Remove tasks from the pool."""
         itasks, bad_items = self.filter_task_proxies(items)
         for itask in itasks:
-            if spawn:
-                self.force_spawn(itask)
             self.remove(itask, 'by request')
         return len(bad_items)
 
-    def trigger_tasks(self, items, back_out=False):
-        """Operator-forced task triggering."""
-        itasks, bad_items = self.filter_task_proxies(items)
-        n_warnings = len(bad_items)
-        for itask in itasks:
+    def trigger_tasks(self, items, back_out=False, task_pool=False):
+        """Forced task triggering.
+
+        task_pool=False:
+           Insert and trigger any abstract task in the workflow.
+        task_pool=True:
+           Match (potentially by state) and trigger existing tasks in the task
+           pool, e.g. to allow retriggering a bunch of unhandled failed tasks.
+
+        """
+        if task_pool:
+            itasks, bad_items = self.filter_task_proxies(items)
+            n_warnings = len(bad_items)
+            for itask in itasks:
+                if back_out:
+                    # (Aborted edit-run, reset for next trigger attempt).
+                    try:
+                        del itask.summary['job_hosts'][itask.submit_num]
+                    except KeyError:
+                        pass
+                    job_d = get_task_job_id(
+                        itask.point, itask.tdef.name, itask.submit_num)
+                    self.job_pool.remove_job(job_d)
+                    itask.submit_num -= 1
+                    itask.summary['submit_num'] = itask.submit_num
+                    itask.local_job_file_path = None
+                    continue
+                if itask.state(*TASK_STATUSES_ACTIVE):
+                    LOG.warning('%s: already triggered' % itask.identity)
+                    n_warnings += 1
+                    continue
+                itask.manual_trigger = True
+                if not itask.state(
+                        TASK_STATUS_QUEUED,
+                        is_held=False
+                ):
+                    itask.state.reset(TASK_STATUS_READY, is_held=False)
+        else:
+            # TODO COMBINE THE FOLLOWING LOGIC WITH THAT OF spawn_tasks?
+            # TODO: edit run back-out for the insert-and-trigger case.
```

Comment:
Trouble is the command line argument forms overlap. Example from one of the several allowed forms:
- trigger any task: `CYCLE-POINT/TASK-NAME-GLOB`
- match and trigger tasks in the pool: `[CYCLE-POINT-GLOB/]TASK-NAME-GLOB[:TASK-STATE]`

If the cycle point is definite (not a glob pattern) and the optional `:STATE` isn't used, we can't tell which is intended.
 
However, I'll put a comment in the code .. it would be nice to clean this up somehow.

Reply:
Leaving this conversation open for now, for the edit-run back-out TODO though (mind you, we have to re-invent edit run in Cylc 8...).